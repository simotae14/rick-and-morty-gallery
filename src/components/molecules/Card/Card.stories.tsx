import type { Meta, StoryObj } from '@storybook/react';

import Card from './index';

const meta = {
  title: 'Molecules/Card',
  component: Card,
  tags: ['autodocs'],
  argTypes: {
    name: {
      control: {
        type: 'text',
      },
    },
    image: {
      control: {
        type: 'text',
      },
    },
    status: {
      control: {
        type: 'text',
      },
    },
    species: {
      control: {
        type: 'text',
      },
    },
  },
  // The default value of the name arg to all stories
  args: { name: 'No Character' },
  parameters: {
    layout: 'centered',
  },
} satisfies Meta<typeof Card>;

export default meta;

type Story = StoryObj<typeof meta>;

export const EmptyProps: Story = {
  args: {
    name: undefined,
  },
};

export const WithProps: Story = {
  args: {
    name: 'Jerry Smith',
    image: 'https://rickandmortyapi.com/api/character/avatar/5.jpeg',
    status: 'Alive',
    species: 'Human',
  },
};
