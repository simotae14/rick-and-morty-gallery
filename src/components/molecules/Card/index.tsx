import CardWrapper, { CardWrapperEnum } from '../../atoms/CardWrapper';
import CardContent from '../../atoms/CardContent';

import { CharacterSmall } from '../../../types/Character.js';
/**
 * Component used to create the cards inside the gallery
 */
function Card({
  name,
  image,
  status,
  species,
  onClickHandler,
}: Partial<Omit<CharacterSmall, 'id'>> & {
  /**
   * Is the action triggered on clicking the Card
   */
  onClickHandler: () => void;
}) {
  return (
    <CardWrapper
      cardWrapperType={CardWrapperEnum.card}
      onClickHandler={onClickHandler}
    >
      <CardContent
        name={name}
        image={image}
        status={status}
        species={species}
      />
    </CardWrapper>
  );
}

export default Card;
