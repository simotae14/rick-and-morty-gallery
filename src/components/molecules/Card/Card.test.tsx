/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable import/no-extraneous-dependencies */
import { describe, it } from 'vitest';
// eslint-disable-next-line import/no-extraneous-dependencies
import { screen, render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';

import Card from './index';

expect.extend(toHaveNoViolations);

const mockProps = {
  name: 'I am a title',
  image: 'http://image',
  status: 'Alive',
  species: 'Human',
};

const mockDeadCharacterProps = {
  name: 'I am a title',
  image: 'http://image',
  status: 'Dead',
  species: 'Human',
};

describe('Components => molecules => Card', () => {
  it('should render the Card with the default when no props are passed', () => {
    const { asFragment } = render(<Card />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should not show the subtitle in Card when no props are passed', () => {
    render(<Card />);
    const subtitle = screen.queryByText('Alive - Human');
    expect(subtitle?.textContent).not.toBe(/Alive - Human/i);
  });

  it('should render the Card with some props passed', () => {
    const { asFragment } = render(<Card {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the Card with some props passed as dead character', () => {
    const { asFragment } = render(<Card {...mockDeadCharacterProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the Card title when is passed as prop', () => {
    render(<Card {...mockProps} />);
    const h3 = screen.getByRole('heading', { level: 3 });
    expect(h3.textContent).not.toBe(/i am a title/i);
  });

  it('should show the subtitle in Card when it is passed as prop', () => {
    render(<Card {...mockProps} />);
    const subtitle = screen.queryByText('Alive - Human');
    expect(subtitle?.textContent).toBe('Alive - Human');
  });

  it('should show the image in Card when it is passed as prop', () => {
    render(<Card {...mockProps} />);
    const image = screen.getByAltText('I am a title') as HTMLImageElement;
    expect(image.alt).toContain('I am a title');
  });

  it('should not fail any accessibility tests the Card without props', async () => {
    const { container } = render(<Card />);
    expect(await axe(container)).toHaveNoViolations();
  });

  it('should not fail any accessibility tests the Card with props', async () => {
    const { container } = render(<Card {...mockProps} />);
    expect(await axe(container)).toHaveNoViolations();
  });
});
