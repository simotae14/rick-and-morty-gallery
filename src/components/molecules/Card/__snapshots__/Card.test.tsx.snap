// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components => molecules => Card > should render the Card with some props passed 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir sc-buuUZy kfzyXq dcAPxy"
  >
    <div
      class="sc-dujCx jdGAeT"
    >
      <img
        alt="I am a title"
        class="sc-bFmjqi YAYvZ"
        src="http://image"
        width="140"
      />
      <h3
        class="sc-iGsdBT gcjGsg"
      >
        I am a title
      </h3>
      <p
        class="sc-ckeRpf flOcDv"
      >
        <span
          class="sc-kRYBTm jkbQkO"
          status="Alive"
        />
        Alive - Human
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Components => molecules => Card > should render the Card with some props passed as dead character 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir sc-buuUZy kfzyXq dcAPxy"
  >
    <div
      class="sc-dujCx jdGAeT"
    >
      <img
        alt="I am a title"
        class="sc-bFmjqi YAYvZ"
        src="http://image"
        width="140"
      />
      <h3
        class="sc-iGsdBT gcjGsg"
      >
        I am a title
      </h3>
      <p
        class="sc-ckeRpf flOcDv"
      >
        <span
          class="sc-kRYBTm jyKKPJ"
          status="Dead"
        />
        Dead - Human
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Components => molecules => Card > should render the Card with the default when no props are passed 1`] = `
<DocumentFragment>
  <div
    class="sc-eLqIir sc-buuUZy kfzyXq dcAPxy"
  >
    <div
      class="sc-dujCx jdGAeT"
    >
      <img
        alt="No Character"
        class="sc-bFmjqi YAYvZ"
        src="../../../assets/rick-morty.svg"
        width="140"
      />
      <h3
        class="sc-iGsdBT gcjGsg"
      >
        No Character
      </h3>
      <p
        class="sc-ckeRpf flOcDv"
      >
        <span
          class="sc-kRYBTm ewBspn"
        />
         - 
      </p>
    </div>
  </div>
</DocumentFragment>
`;
